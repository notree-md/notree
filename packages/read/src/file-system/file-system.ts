import { GraphData } from '@mindgraph/types';
import * as fs from 'fs';
import * as readline from 'node:readline';
import {
  HIDDEN_FILES_REGEX,
  LINK_CONTENT_REGEX,
  MARKDOWN_EXTENSION,
} from '../constants';
import { Provider } from '../types';

export interface FileSystemProviderArgs {
  path: string;
}

export const FileSystem: Provider<FileSystemProviderArgs> = {
  async read({ path }) {
    return build_graph(path);
  },
};

async function build_graph(
  path: string,
  graph: GraphData = { nodes: [], links: [] },
): Promise<GraphData> {
  const dir = await fs.promises.opendir(path);

  for await (const dirent of dir) {
    await add_dirent_to_graph(path, dirent, graph);
  }

  return graph;
}

async function add_dirent_to_graph(
  path: string,
  dirent: fs.Dirent,
  graph: GraphData,
) {
  if (HIDDEN_FILES_REGEX.test(dirent.name)) return;

  const direntPath = `${path}/${dirent.name}`;

  if (dirent.isDirectory()) {
    await build_graph(direntPath, graph);
  } else if (dirent.isFile()) {
    const linkCount = await add_links_to_graph(direntPath, graph);
    graph.nodes.push({
      id: direntPath,
      name: dirent.name,
      linkCount,
    });
  }
}

async function add_links_to_graph(
  filePath: string,
  graph: GraphData,
): Promise<number> {
  let linkCount = 0;

  const fileStream = fs.createReadStream(filePath);
  const lines = readline.createInterface({ input: fileStream });

  for await (const line of lines) {
    const links = line.match(LINK_CONTENT_REGEX) || [];

    for (const link of links) {
      const path = LINK_CONTENT_REGEX.exec(link)?.at(1);

      if (is_valid_link_path(path)) {
        linkCount++;

        const linkDirections = path.split('/');
        const pathToTargetFile = filePath.split('/');

        pathToTargetFile.pop();

        for (const direction of linkDirections) {
          switch (direction) {
            case '.':
              break;
            case '..':
              pathToTargetFile.pop();
              break;
            default:
              pathToTargetFile.push(direction);
          }
        }

        graph.links.push({
          source: filePath,
          target: pathToTargetFile.join('/'),
        });
      }
    }
  }

  return linkCount;
}

function is_valid_link_path(path: string | undefined): path is string {
  return !!(path && !path.includes('://') && path.includes(MARKDOWN_EXTENSION));
}
