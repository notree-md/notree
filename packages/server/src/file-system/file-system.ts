import { GraphData, Node, Link } from '@notree/common';
import * as fs from 'fs';
import * as readline from 'node:readline';
import { HIDDEN_FILES_REGEX, extractLinksFromLine, newNode } from '../common';
import { Provider } from '../types';

type Objects = Record<string, Node | Link>;

export interface FileSystemProviderArgs {
  path: string;
}

export const FileSystem: Provider<FileSystemProviderArgs> = {
  async read({ path }) {
    const objects = await gather_objects_from_directory(path);

    const graphData: GraphData = { nodes: [], links: [] };

    for (const [, item] of Object.entries(objects)) {
      if (!isNode(item)) {
        item.source.childLinks.push(item);
        item.source.childNodes.push(item.target);
        item.target.parentLinks.push(item);
        item.target.parentNodes.push(item.source);

        graphData.links.push(item);
      } else {
        graphData.nodes.push(item);
      }
    }

    return graphData;
  },
};

async function gather_objects_from_directory(
  path: string,
  objects: Objects = {},
): Promise<Objects> {
  const directory = await fs.promises.opendir(path);

  for await (const dirent of directory) {
    if (HIDDEN_FILES_REGEX.test(dirent.name)) continue;

    const direntPath = `${path}/${dirent.name}`;
    if (dirent.isDirectory()) {
      await gather_objects_from_directory(direntPath, objects);
    } else if (dirent.isFile()) {
      objects[direntPath] = newNode({ id: direntPath, title: dirent.name });
    }
  }

  const allNodeIds = Object.keys(objects);

  for (const id of allNodeIds) {
    const item = objects[id];
    if (isNode(item)) {
      await gather_links_from_file(item.id, objects);
    }
  }

  return objects;
}

async function gather_links_from_file(filePath: string, objects: Objects) {
  const fileStream = fs.createReadStream(filePath);
  const lines = readline.createInterface({ input: fileStream });

  for await (const line of lines) {
    const links = extractLinksFromLine(line, filePath);
    for (const link of links) {
      objects[link.source + link.target] = {
        source: objects[link.source] as Node,
        target: objects[link.target] as Node,
      };
    }
  }
}

function isNode(item: Node | Link): item is Node {
  return (item as Node).id !== undefined;
}
