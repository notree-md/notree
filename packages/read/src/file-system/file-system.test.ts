import { describe, expect, it } from 'vitest';
import { FileSystem } from './file-system';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

describe('FileSystem Provider', () => {
  it('should produce the expected graph', async () => {
    const graph = await FileSystem.read({
      path: path.resolve(__dirname, '../../../../', 'draw', 'example', 'notes'),
    });

    expect(graph).toMatchSnapshot();
  });
});
