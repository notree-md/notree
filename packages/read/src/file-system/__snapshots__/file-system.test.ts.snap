// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileSystem Provider > should produce the expected graph 1`] = `
[
  {
    "linkCount": 8,
    "name": "big.md",
  },
  {
    "linkCount": 0,
    "name": "small6.md",
  },
  {
    "linkCount": 0,
    "name": "small2.md",
  },
  {
    "linkCount": 0,
    "name": "small3.md",
  },
  {
    "linkCount": 0,
    "name": "small4.md",
  },
  {
    "linkCount": 0,
    "name": "small5.md",
  },
  {
    "linkCount": 0,
    "name": "small1.md",
  },
  {
    "linkCount": 3,
    "name": "med2.md",
  },
  {
    "linkCount": 3,
    "name": "med1.md",
  },
]
`;
